src/CsrfHeaderMiddleware.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Csrf;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Http\Method;
+use Yiisoft\Http\Status;
+
+use function in_array;
+
+/**
+ * PSR-15 middleware that takes care of custom HTTP header validation.
+ *
+ * @link https://www.php-fig.org/psr/psr-15/
+ * @link https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#employing-custom-request-headers-for-ajaxapi
+ */
+final class CsrfHeaderMiddleware implements MiddlewareInterface
+{
+    public const HEADER_NAME = 'X-CSRF-Header';
+
+    private string $headerName = self::HEADER_NAME;
+    private array $unsafeMethods = [Method::GET, Method::HEAD, Method::POST];
```

Comment:
It may be considered unsafe by default POST, PUT, DELETE, PATCH?

Reply:
unsafe methods are methods without CORS preflight request